okhttp/src/main/java/okhttp3/internal/http2/Http2Stream.kt

```diff
@@ -13,321 +13,291 @@
-      headersQueue.add(headers);
-    }
+  @get:Synchronized internal var errorCode: ErrorCode? = null
 
-    if (isLocallyInitiated() && headers != null) {
-      throw new IllegalStateException("locally-initiated streams shouldn't have headers yet");
-    } else if (!isLocallyInitiated() && headers == null) {
-      throw new IllegalStateException("remotely-initiated streams should have headers");
-    }
-  }
+  /** The exception that explains [errorCode]. Null if no exception was provided. */
+  internal var errorException: IOException? = null
 
-  public int getId() {
-    return id;
+  init {
+    if (headers != null) {
+      check(!isLocallyInitiated) { "locally-initiated streams shouldn't have headers yet" }
+      headersQueue += headers
+    } else {
+      check(isLocallyInitiated) { "remotely-initiated streams should have headers" }
+    }
   }
 
   /**
    * Returns true if this stream is open. A stream is open until either:
    *
-   * <ul>
-   *     <li>A {@code SYN_RESET} frame abnormally terminates the stream.
-   *     <li>Both input and output streams have transmitted all data and headers.
-   * </ul>
+   *  * A `SYN_RESET` frame abnormally terminates the stream.
+   *  * Both input and output streams have transmitted all data and headers.
    *
-   * <p>Note that the input stream may continue to yield data even after a stream reports itself as
+   * Note that the input stream may continue to yield data even after a stream reports itself as
    * not open. This is because input data is buffered.
    */
-  public synchronized boolean isOpen() {
-    if (errorCode != null) {
-      return false;
-    }
-    if ((source.finished || source.closed)
-        && (sink.finished || sink.closed)
-        && hasResponseHeaders) {
-      return false;
+  val isOpen: Boolean
+    @Synchronized get() {
+      if (errorCode != null) {
+        return false
+      }
+      if ((source.finished || source.closed) &&
+          (sink.finished || sink.closed) &&
+          hasResponseHeaders) {
+        return false
+      }
+      return true
     }
-    return true;
-  }
-
-  /** Returns true if this stream was created by this peer. */
-  public boolean isLocallyInitiated() {
-    boolean streamIsClient = ((id & 1) == 1);
-    return connection.client == streamIsClient;
-  }
 
-  public Http2Connection getConnection() {
-    return connection;
-  }
+  /** Returns true if this stream was created by this peer.  */
+  val isLocallyInitiated: Boolean
+    get() {
+      val streamIsClient = (id and 1) == 1
+      return connection.client == streamIsClient
+    }
 
   /**
    * Removes and returns the stream's received response headers, blocking if necessary until headers
    * have been received. If the returned list contains multiple blocks of headers the blocks will be
    * delimited by 'null'.
    */
-  public synchronized Headers takeHeaders() throws IOException {
-    readTimeout.enter();
+  @Synchronized @Throws(IOException::class)
```

Comment:
nit: since it's internal code, you can omit `@Throws(IOException::class)`, unless you actually plan to catch it

Reply:
Not fixed. I want to make the `@Throws` decision once and then comb through the entire codebase.